src/services/db/TokenService.ts

```diff
@@ -0,0 +1,340 @@
+import { ModelStatic } from "sequelize"
+
+import logger from "@logger/logger"
+
+import DatabaseError from "@errors/DatabaseError"
+import NoAvailableTokenError from "@errors/NoAvailableTokenError"
+import TokenParsingError from "@errors/TokenParsingError"
+
+import { AccessToken } from "@database/models"
+
+// Env vars
+export const GITHUB_TOKEN_LIMIT = 5000
+export const GITHUB_TOKEN_THRESHOLD = 4000 // allowed uses
+export const GITHUB_RESET_INTERVAL = 60 * 60 // seconds
+export const ACTIVE_TOKEN_ALERT_1 = 0.6
+export const ACTIVE_TOKEN_ALERT_2 = 0.8
+export const GITHUB_TOKEN_REMAINING_HEADER = "x-ratelimit-remaining"
+export const GITHUB_TOKEN_RESET_HEADER = "x-ratelimit-reset"
+
+export type MaybeResetTime = Result<number, null>
+export const NoResetTime: MaybeResetTime = err(null)
+
+export type TokenData = {
+  id: number
+
+  tokenString: string
+
+  remainingRequests: number
+
+  resetTime: MaybeResetTime
+}
+
+export type MaybeTokenData = Result<TokenData, null>
+export const NoTokenData: MaybeTokenData = err(null)
+
+type ResponseTokenData = {
+  token: string
+  remainingRequests: number
+  resetTime: number
+}
+
+export function queryActiveTokens(
+  tokenDB: ModelStatic<AccessToken>
+): ResultAsync<TokenData[], DatabaseError> {
+  return fromPromise(
+    tokenDB.findAll({
+      where: {
+        isReserved: false,
+      },
+    }),
+    (error) =>
+      new DatabaseError("Unable to retrieve active tokens from database")
+  ).map((activeTokens) =>
+    activeTokens.map((activeToken) => ({
+      id: activeToken.id,
+      tokenString: activeToken.token,
+      remainingRequests: GITHUB_TOKEN_LIMIT,
+      resetTime: NoResetTime,
+    }))
+  )
+}
+
+type LoggerType = typeof logger
+
+export function activeUsageAlert(
+  activeTokensData: TokenData[],
+  useLogger: LoggerType
+) {
+  let exhaustedTokensCount = 0
+  activeTokensData.forEach((tokenData) => {
+    if (
+      tokenData.remainingRequests <=
+      GITHUB_TOKEN_LIMIT - GITHUB_TOKEN_THRESHOLD
+    ) {
+      exhaustedTokensCount += 1
+    }
+  })
+  if (exhaustedTokensCount >= activeTokensData.length * ACTIVE_TOKEN_ALERT_2) {
+    useLogger.info(`${ACTIVE_TOKEN_ALERT_2}% of access token capacity reached`)
+    useLogger.info(
+      `${exhaustedTokensCount}/${activeTokensData.length} active tokens exhausted`
+    )
+  } else if (
+    exhaustedTokensCount >=
+    activeTokensData.length * ACTIVE_TOKEN_ALERT_1
+  ) {
+    useLogger.info(`${ACTIVE_TOKEN_ALERT_1}% of access token capacity reached`)
+    useLogger.info(
+      `${exhaustedTokensCount}/${activeTokensData.length} active tokens exhausted`
+    )
+  }
+}
+
+export function parseResponseTokenData(
+  response: AxiosResponse
+): Result<ResponseTokenData, TokenParsingError> {
+  // response.config.headers.Authorization format: token ghp_********************************
+  if (
+    typeof response.config?.headers?.Authorization !== "string" ||
+    response.config?.headers?.Authorization.length !== 46 ||
```

Comment:
sanity check here: are we sure that github tokens are always a lenght of 46?

Reply:
I am unable to find any documentation that github tokens will always be length 40. However, from the tokens that we currently have in our database and past experiences with using github tokens, I am fairly confident that it always have length 40. If a token of length not 40 is inserted in the database it can be amended.

Reply:
Arh, your flow of thought makes sense, and I agree, I am fairly confident that the length of a github token will prob remain at 40. 
However, when dealing with production code, it might not wise to depend on behaviour defined by external parties when it is not explicitly defined in their documentation unless there is strong reason to do so. I would rather have lesser developer overhead of ensuring that the tokens are of this length when adding them to our db!

Reply:
That's true. Perhaps this check isn't suitable at the moment and I'll drop it until I can get further confirmation that tokens will be length 40.